kernel/aster-nix/src/vm/vmar/vm_mapping.rs

```diff
@@ -227,31 +200,60 @@ impl VmMapping {
         self.inner.lock().is_destroyed
     }
 
+    /// Returns whether the mapping is a shared mapping.
+    pub fn is_shared(&self) -> bool {
+        self.is_shared
+    }
+
     pub fn handle_page_fault(
         &self,
         page_fault_addr: Vaddr,
         not_present: bool,
         write: bool,
     ) -> Result<()> {
-        let vmo_offset = self.vmo_offset() + page_fault_addr - self.map_to_addr();
-        if vmo_offset >= self.vmo.size() {
-            return_errno_with_message!(Errno::EACCES, "page fault addr is not backed up by a vmo");
-        }
-        let page_idx = vmo_offset / PAGE_SIZE;
-        if write {
-            self.vmo.check_rights(Rights::WRITE)?;
-        } else {
-            self.vmo.check_rights(Rights::READ)?;
-        }
-
         let required_perm = if write { VmPerms::WRITE } else { VmPerms::READ };
         self.check_perms(&required_perm)?;
 
-        let frame = self.vmo.get_committed_frame(page_idx, write)?;
+        if !self.is_shared && write && !not_present {
+            // Perform COW at page table.
+            let root_vmar = self.parent.upgrade().unwrap();
+            root_vmar
+                .vm_space()
+                .cow_at(page_fault_addr.align_down(PAGE_SIZE))?;
+            return Ok(());
+        }
+
+        let vmo_offset = self.vmo_offset() + page_fault_addr - self.map_to_addr();
+        let page_idx = vmo_offset / PAGE_SIZE;
 
-        // If read access to cow vmo triggers page fault, the map should be readonly.
+        // If read access to private mapping triggers page fault, the map should be readonly.
         // If user next tries to write to the frame, another page fault will be triggered.
-        let is_readonly = self.vmo.is_cow_vmo() && !write;
+        let mut is_readonly = !self.is_shared && !write;
+
+        let frame = if let Some(vmo) = &self.vmo {
+            if let Ok(page) = vmo.get_committed_frame(page_idx) {
+                if !self.is_shared && write {
+                    // Perform COW directly.
+                    Frame::new_from(&page)?
+                } else {
+                    // Need COW later if the mapping is not shared.
+                    page
+                }
+            } else if !self.is_shared {
+                // The page index is outside the VMO. This is only allowed in private mapping.
+                is_readonly = false;
```

Comment:
Currently, only reading the private and filebacked mapping need to map with read-only flag here. The reason is that:
If a private anonymous mapping commit a page into the page table for reading and you map it with read-only, later you write it ——
- If this process does not perform `fork` between the reading and writing, then this page will actually only be used in this process (its parents will not own this page, otherwise the process does not need to map it), so why need to set the page read-only at first and do an additional copy-on-write？
- If this process performs `fork` between the reading and writing, then the page will be set to read-only during `fork`, so there is also no need to set read-only at first.

If the page index is outside the VMO, we can treat it as private anounymous mapping, so here set `is_readonly` as false here.

The comments are out-of-date, I will update the comments above.